### ISSUE TITLE
Overzealous assumption of a dependency among two tasks if they work on the same directory.

### ISSUE THREAD

**Issue by twwwt** [gradle/gradle#17961]
### Expected Behavior
It there are two tasks, say `A` and `B`, where task `A` produces *output* into a directory `d` and where task `B` uses `d` as an input, Gradle should not necessarily assume task `B` would depend on `A`. Assuming such a dependency merely on the fact that a directory is the output of some task and also the input of another task is overzealous.

For instance, if in fact it is *known* a priori that task `A` writes into `d/a` (meaning that `a` here is a subdirectory of `d`) and `B` reads from `d/b` there is no dependency between A and B, simply because they work on disjoint places; i.e., `A` and `B` may very well execute in any order, even concurrently, in this case (of course assuming that there are no other inputs/outputs of `A` and `B` that would not be disjoint). What is more, this is in fact independent of the fact whether `d` is an input or output of either task. As long as it is known that `A` and `B` work on disjoint files/directories below `d`, `d` could be very well an input *and* an output of both `A` *and* `B`.

Another example would be if it is *known* a priori that `A` only works on files that match the pattern `*.css` and task `B` only works on files that match the pattern `*.js`, while it is not actually known where they are located below `d`. `A` and `B` would equally work on disjoint resources; hence, a dependency should not necessarily be assumed (and execution may proceed in any order even concurrently).

The message/warning that Gradle 7 currently issues in such situations (see below) should be reformulated, at least. In fact, if you want to uphold the current behavior of Gradle 7, it appears to me that it is necessary to provide a way to *explicitly* specify that two tasks are *independent* in case they use the same directory as output/input, respectively. Conversely, you might want to consider abandoning this assumption entirely, meaning that two tasks that use the same directory as output/input respectively are considered independent unless an explicit `dependsOn` is specified. The latter appears cleaner and simpler to me, even if there might be a higher chance that someone shots itself into its feet; but this would be a programmer's error anyway. Am I right that this was in fact the behavior of Gradle until version 7?

### Current Behavior
In the situation described above where `/d` is specified as an input and output of both `A` and `B`, Gradle 7.1.1 issues the following message at runtime:

```
Execution optimizations have been disabled for task ':B' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/d'. Reason: Task ':A' uses this output of task ':B' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/d'. Reason: Task ':B' uses this output of task ':A' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```

### Context
I'm using the [gradle-minify-plugin](https://github.com/gradle-webtools/gradle-minify-plugin) and both tasks `JsMinifyTask` and `CssMinifyTask` to process `*.js` and `*.css` files that are all located below some directory specified as the input and output for both tasks.

### Your Environment
Gradle 7.1.1, Java 11.0.12, Mac OS 11.5.1.

**Comment by wolfs** [comment]
Thank you for reporting!

Looking at the [documentation of the minifier plugin](https://github.com/gradle-webtools/gradle-minify-plugin#minimal-configuration), it actually suggest you use different input and output dirs:
```
minification {
    js { //this: JsMinifyTask
        srcDir = project.file("js")
        dstDir = project.file("build/js")
    }
    css { //this: CssMinifyTask
        srcDir = project.file("css")
        dstDir = project.file("build/css")
    }
}
```

So for your use-case, you are defining the src and dst dirs to be the same? Something like
```
minification {
    js { //this: JsMinifyTask
        srcDir = project.file("src")
        dstDir = project.file("src")
    }
    css { //this: CssMinifyTask
        srcDir = project.file("src")
        dstDir = project.file("src")
    }
}
```

In general, Gradle doesn't really handle the case well when the input and the output directory are the same. Note that it already does stop running tasks in parallel which work on the same output directories. We currently don't have a good solution for this, so we suggest folks to separate the input and output directories for the same task.

The deprecation message asks you to choose an ordering between the two tasks. This can be via `dependsOn` or `mustRunAfter`. So you should either do that or choose separate output directories.

> The latter appears cleaner and simpler to me, even if there might be a higher chance that someone shots itself into its feet

The point of this deprecation message is to make sure you don't shoot yourself in the foot.

**Comment by twwwt** [comment]
> So for your use-case, you are defining the src and dst dirs to be the same?

Yes, indeed. However, the fact that the same directory *should* not be used doesn't mean it *must* not be used. I consider this a perfectly fine use case (provided on *knows* that this will not create conflicts that might result in arbitrary effects). Second, this aspect is actually just a side aspect and it is specific to the plugin. It should not be used to divert attention from the real problem - the problem that I wanted to make you aware of by opening this issue. Let me continue.

This issue shows that Gradle shifts to follow a principle in which it would be ok to get into the way of build script developers, justified by the initially noble goal that this would avoid making mistakes (e.g., not using the same directory as an input and output for the same plugin, or different plugins if their execution order is neither implicit nor explicit). Well, let me say this straight: this is small-man's behavior. If there *exist* use cases in which the situation described above will not cause problems you should not block them. Just because there have been substandard developers that f***ed up their build scripts by not really knowing what they were doing (missing knowledge that two Gradle plugins that work on shared resources *might* create unintended effects if their execution order is arbitrary) is poor justification for such a decision. Only if *no* such use case exists the decision would be justified.

More generally, the deprecation message mentioned above and the fact that you intend to disallow the situation described above in a future Gradle version (>= 8?) is one more instance of a violation of the famous design principle: be liberal in what you accept, be conservative in what you do.

The situation described above *should* be allowed and you *should not* get into the way of a build script developer by preventing him/her from doing this in future versions of Gradle. Don't take the substandard as the new standard - you are making yourself smaller. If you want to insist in the check and the message mentioned above, I hereby propose (i) to rephrase it to be more of a warning (something like "hey, you might have specified a build that might cause problems for you; go rather check this and that"), (ii) that does also not pop up as lifecycle message, and most importantly (iii) do not fail the build in such situations. It's the developer's responsibility to get it's stuff right; you can support him/her in doing so but keep your opinions for yourself as long as nobody explicitly asks for them.

**Comment by twwwt** [comment]
> (i) to rephrase it to be more of a warning (something like "hey, seems you have specified a build that might cause problems for you; go rather check this and that"), (ii) that does also not pop up as lifecycle message

Unrelated to this issue, the idea just jumped into my mind that a "Gradle Linter" should do such things. Having such a tool would be cool. I know that Build Scans already go into this direction but still, I think this would be a great extension to Gradle. Also, this would allow to better separate concerns. I guess many of the checks that are currently implemented in Gradle could be moved into such a dedicated tool and `gradle[w]` could concentrate on being a Gradle build *engine* (i.e., on the execution part).
(I'm just writing this to save this idea. If you equally find this a good idea, I could open a dedicated feature request issue.)

**Comment by wolfs** [comment]
> Yes, indeed. However, the fact that the same directory should not be used doesn't mean it must not be used.
> This issue shows that Gradle shifts to follow a principle in which it would be ok to get into the way of build script developers, justified by the initially noble goal that this would avoid making mistakes

This is not only about novice developers shooting themselves in the foot, such practices have performance implications. For example, the task you described above can't be cacheable. Moreover, Gradle doesn't cache the state of the directory between builds, since it is currently not supported for file system watching.

So the practices you describe make it impossible or much harder for Gradle to do its optimizations. If you want to continue to use the practice, you can also opt-out of Gradle's optimizations by using `Task.doNotTrackState()`.

In general, Gradle wants to make it simpler to write correct and performant tasks and plugins. Therefore, it won't be more liberal in what plugin code it accepts, it will become stricter over time.

Though we do not want to reduce the number of use-cases you can solve with Gradle. As you see yourself, this is a thin line with what we want to achieve. This brings me back to your use-case: Why is it important for your build that the minified and the non-minified files are in the same directory?

**Comment by twwwt** [comment]
@wolfs This is just to inform you that I don't have time during the next days to give a detailed answer, especially to your last question. I'll try to find time on the weekend.

**Comment by twwwt** [comment]
> Why is it important for your build that the minified and the non-minified files are in the same directory?

As simple as it is, it keeps the build simpler. We want minified as well as non-minified files to be siblings (i.e., located in the same folder, with the additional difference that minified files have the classifier `-min` in their name. Both variants are processed afterwards by three more tasks down the pipe (filtering, deflating, see below, and finally packaging).

If we would need to specify another output folder for the minified variants (because a future Version of Gradle would force us to do so), we would inevitably have them temporarily in a different location. This would require us  to bring them back together into the same place (because we want them to be siblings of the non-minified variants), probably by yet another task.

What is more, we apply the same processing pattern to create GZipped variants of both the non- and minified variants, also as siblings (i.e., *.js, *-min.js, *.js.gz, *-min.js.gz are all siblings. The Gzip task also uses the same input and output folder. Finally, this processing pattern is equally applied to minify and Gzip CSS stylesheet files.

Obviously, there would be ways to get around a future limitation of not being allowed to specify the same folder as an input and output. But, why? It would be really just a *limitation*.

Which brings me to come back to these two statements:

> In general, Gradle doesn't really handle the case well when the input and the output directory are the same.

> So the practices you describe make it impossible or much harder for Gradle to do its optimizations.

So is the problem here the practices described, or is the problem the rather naive approach to deciding what optimizations can be applied when? Correct me if I'm wrong, from what I read, I conclude that at least one criteria to disable certain optimizations (most probably caching and concurrent execution) in Gradle is the situation that a task has an input and output that are the same folder; or the constellation described by me at the very beginning of this issue. Sorry to say, this is naive. I see no strategy that is worser than this. Even a simple MRSW strategy would be better. The moment I see such things, I have to think about how coordinating/optimzing concurrent access on shared resources in databases works. Perhaps you should think about an approach that follows how this is essentially done by all database coordination approaches: (i) define the notion of a *changeset* for a task (e.g., the files it is working on), (ii) define the *operations* a task may perform on the elements (files) of a changeset (e.g., read and write), (iii) come up with a *conflict* matrix over those operations (i.e., those operations that may run concurrently and those that don't - that conflict) and (iv) come up with a *protocol* to enforce sequential execution of conflicting operations. In such a model, even the case of a single task that may perform conflicting operations on shared resources concurrently would fit in.

Anyway, let me repeat what the intention of this issue was. A Gradle build that produces warnings like this

```
Execution optimizations have been disabled for task ':B' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/d'. Reason: Task ':A' uses this output of task ':B' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/d'. Reason: Task ':B' uses this output of task ':A' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```

finally completed by the message

```
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
```

should not become incompatible with Gradle 8.
